Refactoring of router tests

nibra opened the pull request:
Refactored router tests
- less assertions per test
- speaking names for test methods
- test JRouter instead of JUri
- use Reflection instead of mocks to test internals

nibra committed:
Refactored router tests
 - less assertions per test
 - speaking names for test methods
 - test JRouter instead of JUri
 - use reflection instead of testing mocks

Refactored JRouterTest
 - less assertions per test
 - speaking names for test methods
 - test JRouter instead of JUri
(cherry picked from commit fc6fdb3)

Refactored JRouterSiteTest
 - less assertions per test
 - speaking names for test methods
 - test JRouter instead of JUri
(cherry picked from commit 98d5849)

Removed 'site' client - it's not feasible to get it work
(cherry picked from commit f08cac1)

Refactored JRouterAdministratorTest
(cherry picked from commit 7d95494)

Cleanup
(cherry picked from commit 5398244)

mbabker commented on tests/unit/suites/libraries/cms/router/JRouterSiteTest.php:
```diff
@@ -417,211 +598,515 @@ public function testParseRawRoute($url, $menubool, $expectedVars, $expectedGloba
-		$uri->delVar('testvar');
-		$this->object->runBuildRawRoute($uri);
-		$this->assertEquals('index.php?option=com_ te?st&testvar=testvalue', $uri->toString());
+		$buildRawRouteMethod = new ReflectionMethod('JRouterSite', 'buildRawRoute');
```
We have a `TestReflection` class that has some of these interactions with the `Reflection` API covered.  You could do something like `TestReflection::invoke($object, 'buildRawRoute', array(&$uri));` then run your assertion.

nibra replied:
Does not work because of the call-by-reference problem. That's why `$method->invokeArgs($object, array(&$uri))` has to be used instead of `$method->invoke($object, &$uri)`.

nibra commented:
Gonna repair this tomorrow.

nibra committed:
Disabled 'JRouter::getInstance()' test - it works only in isolation due to internal static instance cache

nibra committed:
Minor fixes
 - removed unused property
 - avoided side effect with global var
 - used reflection to clear instance cache

nibra commented:
Added the same corrections as in the corresponding pull request at joomla-cms.

nibra committed:
Fixed 'expected to be a reference, value given' problem

wilsonge commented:
I've updated this repo to be upto date with joomla-cms/staging so this code is now in this repo :)
